### PR TITLE
Ab add resources

### PR DESCRIPTION
**Pull request by babicahmed**
- Added networking modules 
- Added readme files

### CODE REVIEW

File: modules/NAT_gateway/main.tf [hunk 1]
```diff
@@ -0,0 +1,71 @@
+terraform {
+  required_providers {
+    azurerm = {
+      source = "hashicorp/azurerm"
+      version = "3.26.0"
+    }
+  }
+}
+provider "azurerm"  {
```
Comment (tpeterson66) [inline modules/NAT_gateway/main.tf:9]: We can remove the provider config from each of the modules.

File: modules/NAT_gateway/main.tf [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+# resources
+# NAT gateway
+# https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/nat_gateway
+resource "azurerm_nat_gateway" "this" {
+  name                    = var.name
+  location                = var.location
+  resource_group_name     = var.resource_group_name
+  sku_name                = var.sku_name
+  idle_timeout_in_minutes = var.idle_timeout_in_minutes
+  zones                   = var.zones
```
Comment (tpeterson66) [inline modules/NAT_gateway/main.tf:60]: We might need a way to check if zones are supported, I would default to null for the variable. Some regions do not support regions yet.

File: modules/application_gateway/main.tf [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+  name                = var.public_ip_name
+  location            = var.location
+  resource_group_name = var.resource_group_name
+  allocation_method   = var.allocation_method
+  sku                 = var.public_ip_sku
+  tags                = var.tags
+}
+# user assigned identity
+# https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/user_assigned_identity
+resource "azurerm_user_assigned_identity" "this" {
```
Comment (tpeterson66) [inline modules/application_gateway/main.tf:95]: Need a dynamic value for this, sometimes, we may need to pass the managed identity vs. creating it. We can do a trick w/ count to only create this if requested.

File: modules/application_gateway/main.tf [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+      backend_http_settings,
+      http_listener,
+      request_routing_rule,
+      backend_address_pool,
+      frontend_port,
+      probe,
+      tags
+    ]
+  }
+  identity {
```
Comment (tpeterson66) [inline modules/application_gateway/main.tf:118]: See comment above for the managed identity

File: modules/network_security_group/main.tf [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+variable "tags" {
+  type        = map(string)
+  description = "A map of the tags to use on the resources that are deployed with this module."
+}
+variable "resource_group_name" {
+  description = "The name of the resource group in which to create the network security group."
+  type        = string
+}
+variable "security_rules" {
+  type        = list(object({
```
Comment (tpeterson66) [inline modules/network_security_group/main.tf:31]: Nice!

File: modules/virtual_network/main.tf [hunk 6]
```diff
@@ -25,6 +36,8 @@ variable "tags" {
   }
 }
 
+# resources
+# https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/virtual_network
 resource "azurerm_virtual_network" "this" {
```
Comment (tpeterson66) [inline modules/virtual_network/main.tf:41]: Need to add DNS in here.

**Review by tpeterson66 — APPROVED** [review]
Here is my feedback, good work!!!

File: modules/NAT_gateway/main.tf (hunk 2 above)
Comment (babicahmed) [reply modules/NAT_gateway/main.tf:60]: Currently terraform Azure is not supporting check for availability zones. We can hard-code currently available regions where availability zones are available and do check against it. How often these things are changing? I would say rarely.

**Pull request merged**